### DISCUSSION TITLE
attribute name is not checked - Attribute syntax should be separate for built-in and custom.

### DISCUSSION THREAD

**Discussion by ghost** [vlang/v#15808] [General]
**V version: 0.3.1 fda39bf** 
**OS: ArchLinux** 

**What did you do?**

```v
[heep]  // nonexistent attribute
struct Foo{}

fn main(){
	f := Foo{}
}
```
**What did you expect to see?**
Compile error that `heep` is unknown attribute.
 
**What did you see instead?**
No error.

**Comment by Delta456** [comment]
We don't check it because people can have custom attributes if I recall correctly.

**Comment by ghost** [comment]
Even if custom attributes are used, there should be a mechanism for the compiler to check them.
How are such misspellings detected?

**Comment by JalonSolov** [comment]
They can't be.  Custom attributes are only handled by User code, not the compiler.  All the compiler _can_ do is ignore them.

**Comment by ghost** [comment]
They should be.

**Comment by JalonSolov** [comment]
How?  If I add my own custom attribute to a file, how does the compiler know about it?  It's just an attribute it doesn't recognize.  I might not even include the module where it's defined/handled.

**Comment by JalonSolov** [comment]
At the _very_ most, I could see adding a special option to the compiler to print out _all_ custom attributes seen.  Perhaps _only_ if there is no other reference to the attribute in the code being compiled.

For example, I could see releasing a module that had custom attributes, and including instructions that said something like "If you want this extra special functionality to work, import module_y instead of module_x."

In other words, it would be completely ignored (as it is now), unless you also included the code that knew what to do with it.

**Comment by ghost** [comment]
At least as far as the built-in attributes are concerned, it should be possible.
User attributes can be done by a different means.

**Comment by ghost** [comment]
There would be no reason why both built-in and user-configured attributes should be the same means.

**Comment by JalonSolov** [comment]
I don't understand... if you use the built-in attributes, they will be handled correctly.  If you don't, they won't.  How is the compiler supposed to know that the misspelled `heep` was really supposed to be `heap`?  Especially if I include my own code that does something different with `heep`?

**Comment by ghost** [comment]
As for the attributes of the user settings, we can think about that afterwards.

**Comment by ghost** [comment]
Then why don't you just issue a `notice`?

**Comment by JalonSolov** [comment]
I would not like to see a `notice` for every custom attribute (and especially not for every built-in one).

**Comment by ghost** [comment]
To suppress those notices, it would be possible to inform the compiler by writing user-set attributes to `.vmod`, for example, as an idea.

**Comment by ghost** [comment]
I don't know how to create user-configured attributes, so I'll just say it as I see fit, but in essence, the compiler should be able to know what attributes the user has defined at the time of compilation, right?
Is it absolutely impossible to create a means to make that possible?

**Comment by ghost** [comment]
@JalonSolov  First I would like to check, do you consider it preferable that the attributes remain unchecked?
Or do you believe that there is nothing that can be done about it because it is technically not currently solvable?

**Comment by JalonSolov** [comment]
Custom attributes can be defined but never used without including extra code.  So the compiler _could_ tell when they were used, but there is no way for it to identify if they're in the code but purposely _not_ used.

As I mentioned above, they may only be used if particular modules are imported, but ignored otherwise.

I don't want to see a `notice` about something like that, where it doesn't affect the code at all, but may not currently be used.

**Comment by ghost** [comment]
So you believe it is preferable to leave the attributes unchecked?

**Comment by ghost** [comment]
Ah, I see.
Are you implying that this issue is not a bug, but should be a feature request?
If so, I guess so.
Then you can change this issue to a feature request.

**Comment by ghost** [comment]
In any case, technical problems need to be overcome so that attributes can be verified.
It's a breeding ground for bugs.

**Comment by shove70** [comment]
From the user's perspective, let's talk about this issue.



What is the purpose of user-defined attributes? It is usually used in reflective behavior. Users want to assign a specific attribute mark to certain structures to perform a unified behavior during reflection.

In an actual production project, it is often used for ORM data entities, Web framework controllers, interceptors (middleware), and so on. The number of them may be very large, and even there will be relevant automatic generators to help generate these structures.

These structures are neatly placed in the project, but it is not guaranteed that every structure will be used. Users have to decide which ones will be used and which ones will not be used in the iterative process.

Therefore, I don't think it is necessary to check whether user attributes are used. As for the tips on spelling errors, I think that's another thing. For example, a special "spelling check" module is added

**Comment by ghost** [comment]
> For example, a special "spelling check" module is added

Positive, constructive conversations are welcome.

The problem I see is that bugs that are hard to find where they are should be mechanically supported as much as possible.

I do not care what means are used to achieve this.

For example, since `heap` has a significant impact on program behavior, the presence or absence of simple mistakes such as misspelling is not a small matter.

Again, my view is that the machine should support avoiding bugs caused by easy mistakes as much as possible.

So this problem may be solved by VLS.

I am not interested in what means will be used to resolve this issue.

What do you think in terms of preventing bugs due to unnecessary easy mistakes? @shove70

**Comment by ghost** [comment]
If necessary for this purpose, it would even be better to fundamentally rethink the design of existing attributes and redesign them from scratch.
Of course, that would be costly, and I know you would not want to make that decision.

So, for the time being, one approach would be to consider the user's custom and built-in attributes separately.

**Comment by shove70** [comment]
It can be understood as a "feature request",

The cost of implementing it is not high, mainly to discuss its necessity.

I still think that some behaviors are "user behaviors", and user behaviors are inexhaustible. If the compiler constantly considers such inexhaustible behaviors, it may become more and more bloated

**Comment by ghost** [comment]
I'm talking about built-in attributes.

**Comment by ghost** [comment]
And I would like to hear your opinion from the point of view of bug detection, not from the point of view of the number of user attributes.

**Comment by ghost** [comment]
As for changing the label, I don't have the privilege to change it, so anyone who has the authority to change it can do so in the feature request.

**Comment by ghost** [comment]
Ah, I understand. What's wrong is that the built-in and custom attributes have the same syntax.
Since they are the same, there is no way to distinguish between them and therefore no way to deal with them.
We just need to change the format so that the compiler can distinguish between built-in and custom attributes.

**Comment by shove70** [comment]
In view of your idea, I think the best way is to add a "spell check" module, which is similar to the relevant functions of Office Word.

The compiler doesn't know whether you want to write [heap] or [heep]. The compiler will respect users, so from the perspective of compiler, it shouldn't notice too much about user behavior.

However, it is good to provide some auxiliary functions like "spell check"

**Comment by ghost** [comment]
And simply as a marker, we can make sure that anything equal to a literal string is no-checked and anything equal to a keyword is checked.
The syntax currently does not make this distinction.

**Comment by shove70** [comment]
> And simply as a marker, we can make sure that anything equal to a literal string is no-checked and anything equal to a keyword is checked. The syntax currently does not make this distinction.

You have a lot of ideas!  👍

**Comment by ghost** [comment]
For example, it is wrong to put attributes related to HttpServer endpoints and attributes related to memory management at the same level.
They should be clearly distinguished and handled separately.

**Comment by ghost** [comment]
For example, how about changing user-defined custom attributes to be written as `@"xxxxxx"`?
Since user-defined attributes are technically just literal strings, enclosing them in double quotes will help convey their meaning.

**Reply by Delta456** [reply]
Do you mean something like `[@custom]` if so then it will look really ugly?

**Reply by ghost** [reply]
No, for example, something like
```v
@"some_attr"
fn foo() {
```

**Reply by ghost** [reply]
Well, whatever the format is, why not something cool one?
Personally, I don't like @ though.

**Reply by ghost** [reply]
I just suggested `@"xxxx"` as an example because it is popular to use @ in various languages, so I personally think anything is fine as long as it can be distinguished from the built-in attributes.
